src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TypeEnter.java

```diff
@@ -1337,25 +1340,76 @@ public Type constructorType() {
-                    params.add(new VarSymbol(PARAMETER | p.flags(), p.name, argtypes.head, csym));
+            // we should do this only once
+            if (constructorSymbol == null) {
+                MethodSymbol csym = super.constructorSymbol();
+                csym.flags_field |= ANONCONSTR | (constr.flags() & VARARGS);
+                csym.flags_field |= based ? ANONCONSTR_BASED : 0;
+
+                csym.appendAttributes(constr.getRawAttributes());
+                csym.appendUniqueTypeAttributes(constr.getRawTypeAttributes().stream()
+                        .filter(anno -> anno.position.type == TargetType.METHOD_RETURN)
+                        .collect(List.collector()));
+
+                ListBuffer<VarSymbol> params = new ListBuffer<>();
+                List<Type> argtypes = constructorType().getParameterTypes();
+                if (!enclosingType().hasTag(NONE)) {
                     argtypes = argtypes.tail;
+                    params = params.prepend(new VarSymbol(PARAMETER, make.paramName(0), enclosingType(), csym));
+                }
+                if (constr.params != null) {
+                    for (VarSymbol p : constr.params) {
+                        VarSymbol param = new VarSymbol(PARAMETER | p.flags(), p.name, argtypes.head, csym);
+                        param.appendAttributes(p.getRawAttributes());
+                        param.appendUniqueTypeAttributes(p.getRawTypeAttributes());
+                        params.add(param);
+                        argtypes = argtypes.tail;
+                    }
                 }
+                csym.params = params.toList();
+
+                // now lets create the tree to make sure that it is created according to our needs
+                methodDecl = make.MethodDef(constrParams(csym),
+                        typeParams(csym.type.getTypeArguments()), csym, csym.type, null);
+                return csym;
             }
-            csym.params = params.toList();
-            return csym;
+            return constructorSymbol;
         }
 
+        // helper methods
+            private List<JCVariableDecl> constrParams(MethodSymbol csym) {
+                ListBuffer<JCVariableDecl> params = new ListBuffer<>();
+                if (csym.params != null) {
+                    for (VarSymbol param : csym.params) {
+                        /* the compiler will issue an error if a type annotation is applied to, for example, `java.lang.String`
+                         * but it won't if the type annotation is applied to `String`, at this point all types will be represented
+                         * in its fully unfolded form so we need to go back to the short form if type annotations are present for a
+                         * given parameter
+                         */
+                        if (!param.getRawTypeAttributes().isEmpty()) {
+                            params.append(make.VarDef(param, make.Ident(param.type.tsym), null));
```

Comment:
I guess I would only use this branch, whether or not there is a type annotation. (Would be much better tested, among other things.)